staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go

```diff
@@ -0,0 +1,345 @@
+}
+
+// listMap collects information about a single list with map semantics.
+type listMap struct {
+	declaredAsMap bool
+	keyFields     []string
+}
+
+type listTypeTagValidator struct {
+	byFieldPath map[string]*listMap
+}
+
+func (listTypeTagValidator) Init(Config) {}
+
+func (listTypeTagValidator) TagName() string {
+	return listTypeTag
+}
+
+func (listTypeTagValidator) ValidScopes() sets.Set[Scope] {
+	return listTagsValidScopes
+}
+
+func (lttv listTypeTagValidator) GetValidations(context Context, _ []string, payload string) (Validations, error) {
+	t := context.Type
+	if t.Kind == types.Alias {
+		t = t.Underlying
+	}
+	if t.Kind != types.Slice && t.Kind != types.Array {
+		return Validations{}, fmt.Errorf("can only be used on list types")
+	}
+
+	if payload == "map" {
+		if realType(t.Elem).Kind != types.Struct {
+			return Validations{}, fmt.Errorf("only lists of structs can be list-maps")
+		}
+
+		// Save the fact that this list is a map.
+		if lttv.byFieldPath[context.Path.String()] == nil {
+			lttv.byFieldPath[context.Path.String()] = &listMap{}
+		}
+		lm := lttv.byFieldPath[context.Path.String()]
+		lm.declaredAsMap = true
+	} // FIXME:else?
+
+	// This tag doesn't generate any validations.  It just accumulates
+	// information for other tags to use.
+	return Validations{}, nil
+}
+
+func realType(t *types.Type) *types.Type {
+	for {
+		if t.Kind == types.Alias {
+			t = t.Underlying
+		} else if t.Kind == types.Pointer {
+			t = t.Elem
+		} else {
+			break
+		}
+	}
+	return t
+}
+
+func (lttv listTypeTagValidator) Docs() TagDoc {
+	doc := TagDoc{
+		Tag:         lttv.TagName(),
+		Scopes:      lttv.ValidScopes().UnsortedList(),
+		Description: "Declares a list field's semantic type.",
+		Payloads: []TagPayloadDoc{{
+			Description: "<type>",
+			Docs:        "map | atomic",
+		}},
+	}
+	return doc
+}
+
+type listMapKeyTagValidator struct {
+	byFieldPath map[string]*listMap
+}
+
+func (listMapKeyTagValidator) Init(Config) {}
+
+func (listMapKeyTagValidator) TagName() string {
+	return listMapKeyTag
+}
+
+func (listMapKeyTagValidator) ValidScopes() sets.Set[Scope] {
+	return listTagsValidScopes
+}
+
+func (lmktv listMapKeyTagValidator) GetValidations(context Context, _ []string, payload string) (Validations, error) {
+	t := context.Type
+	if t.Kind == types.Alias {
+		t = t.Underlying
+	}
+	if t.Kind != types.Slice && t.Kind != types.Array {
+		return Validations{}, fmt.Errorf("can only be used on list types")
+	}
+	if realType(t.Elem).Kind != types.Struct {
+		return Validations{}, fmt.Errorf("only lists of structs can be list-maps")
+	}
+
+	found := false
+	for _, memb := range realType(t.Elem).Members {
+		//FIXME: should this be the JSON name?  That's a change...
+		if memb.Name == payload {
+			if k := realType(memb.Type).Kind; k != types.Builtin {
+				return Validations{}, fmt.Errorf("only primitive types can be list-map keys, not %s", k)
+			}
+			found = true
+			break
+		}
```

Comment:
Yes, we should use JSON name here. This will become important later when we deal with OpenAPI publication.

Reply:
all the existing keys are defined as Go names -- should I just try to use either one?  If I can't find a JSON name, use the Go?  Or maybe just make sure it hard-fails, so authors need to do it right?

Reply:
Changed it to just error if the JSON name is not found.